go/tools/builders/compilepkg.go

```diff
@@ -383,16 +386,24 @@ func compileArchive(
 		}
 	}
 
+	pkgDefFile := filepath.Join(filepath.Dir(outPath), pkgDef)
+	if err = extractPkgDef(outPath, pkgDefFile); err != nil {
+		return fmt.Errorf("failed to extract %s from %s: %v", pkgDef, outPath, err)
+	}
 	// Check results from nogo.
 	if nogoChan != nil {
 		err := <-nogoChan
 		nogoChan = nil // no cancellation needed
 		if err != nil {
 			return err
 		}
+		// TODO is it possible for factOut not created in this case?
```

Comment:
`nogo` should write the `-x` file if it succeeded without errors or diagnostics.

It's hard to read the code and understand whether that happened. We could check `nogoOutput.Len()`, but that doesn't seem readable. Instead, let's change `nogoFailed` to a variable of an enum type, then use that for control flow later on.

```
type nogoResult int

const (
  nogoNotRun nogoResult = iota
  nogoError
  nogoFailed
  nogoSucceeded
)
```